src/olympia/shelves/models.py

```diff
@@ -1,22 +1,28 @@
+from collections import namedtuple
+from urllib import parse
+
 from django.db import models
+from django.utils.functional import cached_property
 
 from olympia import amo
 from olympia.amo.models import ModelBase
-
-ENDPOINTS = ('collections', 'search')
-
-ENDPOINT_CHOICES = tuple((ty, ty) for ty in ENDPOINTS)
+from olympia.tags.models import Tag
 
 
 class Shelf(ModelBase):
+    Endpoints = namedtuple('Endpoints', ['COLLECTIONS', 'SEARCH', 'TAGS'])(
+        'collections', 'search', 'tags'
+    )
+    ENDPOINT_CHOICES = tuple((endpoint, endpoint) for endpoint in Endpoints)
+
     title = models.CharField(max_length=70, help_text='Will be translated.')
     endpoint = models.CharField(
         max_length=200, choices=ENDPOINT_CHOICES, db_column='shelf_type'
     )
     criteria = models.CharField(
         max_length=200,
         help_text='e.g., "?promoted=recommended&sort=random&type=extension" '
-        'or the collection slug',
+        'or the collection slug. For non-collection shelves, for no criteria enter "?"',
```

Comment:
Should it be necessary for a user to enter `?` for no criteria? Couldn't we deal with that for them if they leave it blank?

Reply:
it's possible... but it makes the validation tricky (especially at a database level) because we require a value for collection shelves.  I'll look into it.